### PR TITLE
bsr_dense_bmm(): enable more precise float32 support with float64 accumulators

### PR DESCRIPTION
**Pull request by nikitaved**
Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101163
* __->__ #100882

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
The failure unfortunately looks related. It's surprising because the accumulation precision here is higher and thus we'd expect the results to be closer not further away. We might need to adjust tolerances here.

**Comment by nikitaved** [comment]
The result is very stange. It has the same error as with `float32` accumulators...

File: torch/sparse/_triton_ops.py [hunk 1]
```diff
@@ -195,7 +196,7 @@ def _bsr_strided_dense_rowspace_kernel(
             + col_indices_stride * nnz_offset
         )
 
-        output_acc_block = tl.zeros((BLOCKSIZE_ROW, BLOCKSIZE_ROW), tl.float32)
+        output_acc_block = tl.zeros((BLOCKSIZE_ROW, BLOCKSIZE_ROW), dtype=acc_dtype)
         for _ in range(row_nnz):
             values_block = tl.load(values_block_ptrs)
```
Comment (amjames) [inline torch/sparse/_triton_ops.py:202]: It won't let me add a suggestion outside of the diff, but playing with this locally the tests pass if you modify the `tl.dot` call found a few lines below:

```python
output_acc_block += tl.dot(values_block, dense_block, allow_tf32=False)
```

You can make toggling the flag conditional on the dtype in play, but as far as I know it should only have an effect for float32. 

Note that things do not quite work out if we set that flag to match the value of `torch.backends.cuda.matmul.allow_tf32`. When torch has this flag set true and the triton flag is also set true the results do not match, and the errors are too large for adjusting the dtype to account for them (max ~10^-1 absolute error).

Comment (nikitaved) [reply torch/sparse/_triton_ops.py:203]: Thank you! I will play with it. I wonder whether it affects all mm-like methods...

**Pull request closed without merging**